### ISSUE TITLE
5.8.2: sort CORE::keys() results in subroutine in sort

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6917]
Migrated from [rt.perl.org#24482](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24482) (status was 'resolved')

Searchable as RT24482$

**Comment by p5pRT** [comment]
### From muaddib@ilwrath.mesas.com

#### Created by muaddib@ilwrath.mesas.com



 $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
 Undefined subroutine in sort at \-e line 1\.

 $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
 a

I have verified that this is a problem in 5\.6\.x through present\.  5\.5\.x
does not recognized "CORE&#8203;::keys\(\)"\.  \(I'm not sure how to search to see
if this is a known problem; it's not listed any FAQ\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.2:

Configured by muaddib at Fri Nov  7 15:01:40 CST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=solaris, osvers=2.7, archname=sun4-solaris
    uname='sunos ilwrath.mesas.com 5.7 generic sun4u sparc sunw,ultra-250 '
    config_args='-d -e -s -Dprefix=/mesa/devtrees/myarch -Dscriptdir=/mesa/devtrees/myarch/bin -Dlocincpth=/mesa/devtrees/myarch/include /mesa/myarch/include /usr/gnu/include -Dloclibpth=/mesa/devtrees/myarch/lib /mesa/myarch/lib /usr/gnu/lib -Daltmake=make -Uinstallusrbinperl -Doptimize=-g -Duseshrplib -DDEBUGGING -Dcc=gcc'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/mesa/devtrees/myarch/include -I/mesa/myarch/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/mesa/devtrees/myarch/include -I/mesa/myarch/include'
    ccversion='', gccversion='3.3.2', gccosandvers='solaris2.7'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/mesa/devtrees/myarch/lib -L/mesa/myarch/lib '
    libpth=/mesa/devtrees/myarch/lib /mesa/myarch/lib /usr/lib /usr/ccs/lib /usr/local/lib
    libs=-lsocket -lnsl -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='  -R /mesa/devtrees/myarch/lib/perl5/5.8.2/sun4-solaris/CORE'
    cccdlflags='-fPIC', lddlflags='-G -L/mesa/devtrees/myarch/lib -L/mesa/myarch/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /mesa/devtrees/muaddib/mainline/lib/perl/5.00502
    /mesa/devtrees/muaddib/mainline/lib/perl
    /mesa/devtrees/muaddib/mainline/common/lib/perl
    /mesa/devtrees/myarch/lib/perl5/5.8.2/sun4-solaris
    /mesa/devtrees/myarch/lib/perl5/5.8.2
    /mesa/devtrees/myarch/lib/perl5/site_perl/5.8.2/sun4-solaris
    /mesa/devtrees/myarch/lib/perl5/site_perl/5.8.2
    /mesa/devtrees/myarch/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    HOME=/home/muaddib
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/mesa/oracle/8.1.7.4/lib:/mesa/devtrees/muaddib/mainline/lib:/mesa/devtrees/myarch/lib:/mesa/myarch/lib:/mesa/myarch/tools/dev/lib:/mesa/tools/arch/common/lib:/mesa/tools/arch/sun4_solaris/lib:.:/usr/krb5/lib:/usr/local/lib:/usr/share/lib:/usr/java/lib:/usr/lib:/lib:/usr/X/lib:/usr/ccs/lib
    LOGDIR (unset)
    PATH=/mesa/oracle/8.1.7.4/bin:/mesa/devtrees/muaddib/mainline/sbin:/mesa/devtrees/muaddib/mainline/bin:/mesa/devtrees/myarch/bin:/mesa/bin:/mesa/myarch/tools/dev/bin:/mesa/tools/arch/common/bin:/mesa/tools/arch/sun4_solaris/bin:/home/muaddib/bin/hosts:/home/muaddib/bin/Mail:/opt/share/bin:/usr/bin/X11:/home/muaddib/bin:/home/muaddib/arch/sun4_solaris/bin:/usr/krb5/sbin:/usr/krb5/bin:/mesa/bin:/usr/local/sbin:/usr/local/bin:/usr/java/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/X/bin:/usr/ccs/bin:/usr/bin/mh:/usr/games
    PERLLIB=/mesa/devtrees/muaddib/mainline/lib/perl/5.00502:/mesa/devtrees/muaddib/mainline/lib/perl:/mesa/devtrees/muaddib/mainline/common/lib/perl
    PERL_BADLANG (unset)
    PERL_READLINE_NOWARN=1
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

James Jurach \(via RT\) wrote&#8203;:
> 
>  $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
>  Undefined subroutine in sort at \-e line 1\.
> 
>  $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
>  a

This is due to two facts :

\- perl will parse "sort foo\(\.\.\.\)" and "sort\(foo \.\.\.\)" as "sort foo \.\.\."
&nbsp;  In other words\, it doesn't take the parens into account
\- perl parses CORE&#8203;::keyword as a subroutine\, not as the corresponding
&nbsp;  keyword\.

Fixing this would probably break backward incompatibility in
undesirable ways\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'stalled'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Nov 16 09&#8203;:10&#8203;:38 2003\, rafael wrote&#8203;:
> James Jurach \(via RT\) wrote&#8203;:
> > 
> >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
> >  Undefined subroutine in sort at \-e line 1\.
> > 
> >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
> >  a
> 
> This is due to two facts :
> 
> \- perl will parse "sort foo\(\.\.\.\)" and "sort\(foo \.\.\.\)" as "sort foo \.\.\."
>   In other words\, it doesn't take the parens into account
> \- perl parses CORE&#8203;::keyword as a subroutine\, not as the corresponding
>   keyword\.
> 
> Fixing this would probably break backward incompatibility in
> undesirable ways\.

How would that break compatibility?  Who is currently putting custom
subroutines in the CORE&#8203;:: namespace?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Jan 10 13&#8203;:42&#8203;:33 2012\, sprout wrote&#8203;:
> On Sun Nov 16 09&#8203;:10&#8203;:38 2003\, rafael wrote&#8203;:
> > James Jurach \(via RT\) wrote&#8203;:
> > > 
> > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
> > >  Undefined subroutine in sort at \-e line 1\.
> > > 
> > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
> > >  a
> > 
> > This is due to two facts :
> > 
> > \- perl will parse "sort foo\(\.\.\.\)" and "sort\(foo \.\.\.\)" as "sort foo \.\.\."
> >   In other words\, it doesn't take the parens into account
> > \- perl parses CORE&#8203;::keyword as a subroutine\, not as the corresponding
> >   keyword\.
> > 
> > Fixing this would probably break backward incompatibility in
> > undesirable ways\.
> 
> How would that break compatibility?  Who is currently putting custom
> subroutines in the CORE&#8203;:: namespace?
> 

There's been no correspondence in this ticket for a year\.  Reviewing
this older ticket tonight\, I don't see a strong case that there's a bug
in the source code which needs fixing\.

Anyone disagree?  If not\, I will close the ticket in seven days and am
taking it for that purpose\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Jan 27 17&#8203;:42&#8203;:43 2013\, jkeenan wrote&#8203;:
> On Tue Jan 10 13&#8203;:42&#8203;:33 2012\, sprout wrote&#8203;:
> > On Sun Nov 16 09&#8203;:10&#8203;:38 2003\, rafael wrote&#8203;:
> > > James Jurach \(via RT\) wrote&#8203;:
> > > > 
> > > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
> > > >  Undefined subroutine in sort at \-e line 1\.
> > > > 
> > > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
> > > >  a
> > > 
> > > This is due to two facts :
> > > 
> > > \- perl will parse "sort foo\(\.\.\.\)" and "sort\(foo \.\.\.\)" as "sort foo
\.\.\."
> > >   In other words\, it doesn't take the parens into account
> > > \- perl parses CORE&#8203;::keyword as a subroutine\, not as the corresponding
> > >   keyword\.
> > > 
> > > Fixing this would probably break backward incompatibility in
> > > undesirable ways\.
> > 
> > How would that break compatibility?  Who is currently putting custom
> > subroutines in the CORE&#8203;:: namespace?
> > 
> 
> There's been no correspondence in this ticket for a year\.  Reviewing
> this older ticket tonight\, I don't see a strong case that there's a bug
> in the source code which needs fixing\.
> 
> Anyone disagree?  If not\, I will close the ticket in seven days and am
> taking it for that purpose\.
> 

No one spoke up to argue for the existence of a real bug\.  Hence\, I am
closing this ticket as per schedule\.  Marking ticket Rejected\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 03 05&#8203;:45&#8203;:38 2013\, jkeenan wrote&#8203;:
> On Sun Jan 27 17&#8203;:42&#8203;:43 2013\, jkeenan wrote&#8203;:
> > On Tue Jan 10 13&#8203;:42&#8203;:33 2012\, sprout wrote&#8203;:
> > > On Sun Nov 16 09&#8203;:10&#8203;:38 2003\, rafael wrote&#8203;:
> > > > James Jurach \(via RT\) wrote&#8203;:
> > > > > 
> > > > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort CORE&#8203;::keys\(%a\)\)\,"\\n"\)'
> > > > >  Undefined subroutine in sort at \-e line 1\.
> > > > > 
> > > > >  $ perl \-we '%a=\(a=>"b"\); print\(\(sort keys\(%a\)\)\,"\\n"\)'
> > > > >  a
> > > > 
> > > > This is due to two facts :
> > > > 
> > > > \- perl will parse "sort foo\(\.\.\.\)" and "sort\(foo \.\.\.\)" as "sort foo
> \.\.\."
> > > >   In other words\, it doesn't take the parens into account
> > > > \- perl parses CORE&#8203;::keyword as a subroutine\, not as the
corresponding
> > > >   keyword\.
> > > > 
> > > > Fixing this would probably break backward incompatibility in
> > > > undesirable ways\.
> > > 
> > > How would that break compatibility?  Who is currently putting custom
> > > subroutines in the CORE&#8203;:: namespace?
> > > 
> > 
> > There's been no correspondence in this ticket for a year\.  Reviewing
> > this older ticket tonight\, I don't see a strong case that there's a bug
> > in the source code which needs fixing\.
> > 
> > Anyone disagree?  If not\, I will close the ticket in seven days and am
> > taking it for that purpose\.
> > 
> 
> No one spoke up to argue for the existence of a real bug\.  Hence\, I am
> closing this ticket as per schedule\.  Marking ticket Rejected\.

Actually\, I think this \*is\* a bug\, as CORE&#8203;:: is supposed to disambiguate
and indicate a built\-in keyword\.

sort CORE&#8203;::lc\($\_\) should therefore do the same thing as sort lc\($\_\)\, and
treat ‘CORE&#8203;::lc’ as a keyword\, rather than a sort sub name\.

require happens to do the same thing as sort\.  I consider this a bug\,
too \(the second one\-liner should behave the same as the first\)&#8203;:

$ \./perl \-Ilib \-e 'require lc "FOO\.pl"'
Can't locate foo\.pl in @&#8203;INC \(@&#8203;INC contains&#8203;: lib
/usr/local/lib/perl5/site\_perl/5\.19\.1/darwin\-2level
/usr/local/lib/perl5/site\_perl/5\.19\.1
/usr/local/lib/perl5/5\.19\.1/darwin\-2level /usr/local/lib/perl5/5\.19\.1 \.\)
at \-e line 1\.
$ \./perl \-Ilib \-e 'require CORE&#8203;::lc "FOO\.pl"'
String found where operator expected at \-e line 1\, near "require
CORE&#8203;::lc "FOO\.pl""
&nbsp;    \(Do you need to predeclare require?\)
syntax error at \-e line 1\, near "require CORE&#8203;::lc "FOO\.pl""
Execution of \-e aborted due to compilation errors\.

Fixing this is trivial\.  Are there any objectors?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'rejected' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-04T23&#8203;:37&#8203;:36\]
> Actually\, I think this \*is\* a bug\, as CORE&#8203;:: is supposed to disambiguate
> and indicate a built\-in keyword\.
> 
> sort CORE&#8203;::lc\($\_\) should therefore do the same thing as sort lc\($\_\)\, and
> treat ‘CORE&#8203;::lc’ as a keyword\, rather than a sort sub name\.
> 
> require happens to do the same thing as sort\.  I consider this a bug\,
> too \(the second one\-liner should behave the same as the first\)&#8203;:
> \[\.\.\.\]
> Fixing this is trivial\.  Are there any objectors?

Until somebody comes up with a good specific objection\, I agree with you\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Jun 07 07&#8203;:45&#8203;:54 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-
> 04T23&#8203;:37&#8203;:36\]
> > Actually\, I think this \*is\* a bug\, as CORE&#8203;:: is supposed to
> disambiguate
> > and indicate a built\-in keyword\.
> >
> > sort CORE&#8203;::lc\($\_\) should therefore do the same thing as sort lc\($\_\)\,
> and
> > treat ‘CORE&#8203;::lc’ as a keyword\, rather than a sort sub name\.
> >
> > require happens to do the same thing as sort\.  I consider this a
> bug\,
> > too \(the second one\-liner should behave the same as the first\)&#8203;:
> > \[\.\.\.\]
> > Fixing this is trivial\.  Are there any objectors?
> 
> Until somebody comes up with a good specific objection\, I agree with
> you\.
> 

I have just fixed it in commit 01b5ef5\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
